File: modules/platforms/dotnet/Apache.Ignite/Option.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache.Ignite;

using System;
using System.Diagnostics.CodeAnalysis;
using System.Text;

/// <summary>
/// A wrapper that may or may not contain a value of type <typeparamref name="T"/>.
/// </summary>
/// <typeparam name="T">Value type.</typeparam>
[SuppressMessage("Microsoft.Naming", "CA1716:IdentifiersShouldNotMatchKeywords", Justification = "Reviewed.")]
public readonly record struct Option<T>
{
    private readonly T _value;

    /// <summary>
    /// Initializes a new instance of the <see cref="Option{T}"/> struct.
    /// </summary>
    /// <param name="value">Value.</param>
    /// <param name="hasValue">Whether the value is present.</param>
    [SuppressMessage(
        "StyleCop.CSharp.DocumentationRules",
        "SA1642:ConstructorSummaryDocumentationMustBeginWithStandardText",
        Justification = "False positive.")]
    internal Option(T value, bool hasValue)
    {
        _value = value;
        HasValue = hasValue;
    }

    /// <summary>
    /// Gets the value.
    /// </summary>
    public T Value => HasValue
        ? _value
        : throw new InvalidOperationException("Value is not present. Check HasValue property before accessing Value.");

    /// <summary>
    /// Gets a value indicating whether the value is present.
    /// </summary>
    public bool HasValue { get; }

    /// <summary>
    /// Deconstructs this instance.
    /// </summary>
    /// <param name="value">Value.</param>
    /// <param name="hasValue">Whether the value is present.</param>
    public void Deconstruct(out T value, out bool hasValue)
    {
        value = _value;
        hasValue = HasValue;
    }

    /// <summary>
    /// Maps this instance to another type.
    /// </summary>
    /// <param name="selector">Selector.</param>
    /// <typeparam name="TRes">Result type.</typeparam>
    /// <returns>Resulting option.</returns>
    public Option<TRes> Select<TRes>(Func<T, TRes> selector) => HasValue && selector != null! ? Option.Some(selector(_value)) : default!;

    private bool PrintMembers(StringBuilder builder)
    {
        builder.Append("HasValue = ");
        builder.Append(HasValue);

        if (HasValue)
        {
            builder.Append(", Value = ");
            builder.Append(_value);
        }

        return true;
    }
}

/// <summary>
/// Static helpers for <see cref="Option{T}"/>.
/// </summary>
[SuppressMessage("Microsoft.Naming", "CA1716:IdentifiersShouldNotMatchKeywords", Justification = "Reviewed.")]
public static class Option
{
    /// <summary>
    /// Returns an option of the specified value.
    /// </summary>
    /// <param name="val">Value.</param>
    /// <typeparam name="T">value type.</typeparam>
    /// <returns>Option of T.</returns>
    public static Option<T> Some<T>(T val) => new(val, true);

    /// <summary>
    /// Returns an option without a value.
    /// </summary>
    /// <typeparam name="T">value type.</typeparam>
    /// <returns>Option of T.</returns>
    public static Option<T> None<T>() => default;
}

File: modules/catalog-dsl/src/main/java/org/apache/ignite/internal/catalog/sql/Option.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.internal.catalog.sql;

class Option extends QueryPart {
    private final String name;

    private final Object value;

    private Option(String name, Object value) {
        this.name = name;
        this.value = value;
    }

    public static Option partitions(Integer partitions) {
        return new Option("PARTITIONS", partitions);
    }

    public static Option replicas(Integer replicas) {
        return new Option("REPLICAS", replicas);
    }

    public static Option quorumSize(Integer quorumSize) {
        return new Option("QUORUM_SIZE", quorumSize);
    }

    public static Option dataNodesAutoAdjustScaleUp(Integer adjust) {
        return new Option("DATA_NODES_AUTO_ADJUST_SCALE_UP", adjust);
    }

    public static Option dataNodesAutoAdjustScaleDown(Integer adjust) {
        return new Option("DATA_NODES_AUTO_ADJUST_SCALE_DOWN", adjust);
    }

    public static Option distributionAlgorithm(String distributionAlgorithm) {
        return new Option("DISTRIBUTION_ALGORITHM", distributionAlgorithm);
    }

    public static Option dataRegion(String dataRegion) {
        return new Option("DATAREGION", dataRegion);
    }

    public static Option storageProfiles(String storageProfiles) {
        return new Option("STORAGE_PROFILES", storageProfiles);
    }

    public static Option consistencyMode(String consistencyMode) {
        return new Option("CONSISTENCY_MODE", consistencyMode);
    }

    public static Option name(String name) {
        return new Option("NAME", name);
    }

    public static Option tableName(String tableName) {
        return new Option("TABLE_NAME", tableName);
    }

    public static Option schemaName(String tableName) {
        return new Option("SCHEMA_NAME", tableName);
    }

    public static Option zoneName(String zoneName) {
        return new Option("ZONE_NAME", zoneName);
    }

    public static Option indexId(int indexId) {
        return new Option("INDEX_ID", indexId);
    }

    public static Option filter(String filter) {
        return new Option("DATA_NODES_FILTER", filter);
    }

    @Override
    protected void accept(QueryContext ctx) {
        ctx.sql(name).sql("=");
        boolean isStringValue = value instanceof String;
        if (isStringValue) {
            String strValue = value.toString();
            char quoteChar = '\'';
            if (!QueryUtils.isQuoted(strValue, quoteChar)) {
                ctx.sql(quoteChar).sql(strValue).sql(quoteChar);
            }
        } else {
            ctx.sql(value.toString());
        }
    }
}
